Quality of life improvements:

spillner opened the pull request:
- implemented info-* commands for all package states, not just -new
  - kernel-clean now removes all but the latest installed version
    of kernel packages, rather than any version not matching the repo
    (which risked inadvertently leaving zero versions installed)
  - upgrade no longer touches kernel packages (must use kernel-upgrade)
  - list-kernels shows packages that would be affected by kernel-clean or kernel-upgrade
  - foreign-upgradable and frozen-upgradable show packages now available in newer versions in the repo
  - do not unnecessarily prompt the user when there's no *.new file to compare
  - added aliases for multiple commands to ease transition from apt and similar tools

Commit:
Quality of life improvements:
  - implemented info-* commands for all package states, not just -new
  - kernel-clean now removes all but the latest installed version
    of kernel packages, rather than any version not matching the repo
    (which risked inadvertently leaving zero versions installed)
  - upgrade no longer touches kernel packages (must use kernel-upgrade)
  - list-kernels shows packages that would be affected by kernel-clean or kernel-upgrade
  - foreign-upgradable and frozen-upgradable show packages now available in newer versions in the repo
  - do not unnecessarily prompt the user when there's no *.new file to compare
  - added aliases for multiple commands to ease transition from apt and similar tools
  [Rebased 14 Jan 2024 against upstream v52 master]

spillner committed:
Fixed bug in printing frozen/foreign kernel packages, clarified variable names to reduce future errors
[Rebased 14 Jan 2024 against upstream v52 master]

spillner committed:
Clarified code, output message for upgradable kernels in list-kernels
[Rebased 14 Jan 2024 against upstream v52 master]

spillner committed:
Improved foreign-upgradable/frozen-upgradable to show installed and available versions.
Clarified the section headings in list-kernels.
[Rebased 14 Jan 2024 against upstream v52 master]

spillner committed:
Included new commands (since v48) in the "help all" output.

spillner committed:
Use build tag to differentiate between different packages with the same base version number.
[Rebased 14 Jan 2024 against upstream v52 master]

spillner committed:
Prevent kernel packages from inadvertently showing up in both "up to date" and "close to repo" lists
[Rebased 14 Jan 2024 against upstream v52 master]

aclemons commented:
Sorry for the delay here. I will have a look at these change soon.

spillner commented:
No worries.  I've been busy lately but will take a look at Python3
conversion (or Python 2/3 dual support) sometime soon, which will most
likely be a bigger patch.

Brent

On Tue, Feb 8, 2022 at 2:09 AM Andrew Clemons ***@***.***> wrote:
>
> Sorry for the delay here. I will have a look at these changes soon.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android.
> You are receiving this because you authored the thread.Message ID: ***@***.***>

spillner committed:
Corrected new print statements to use Python 3-compatible syntax.

spillner committed:
Allow access to 'help' command without requiring root privileges

spillner committed:
BUGFIX: remove extraneous parentheses from use of gnupg_exec_name, avoiding 'str object is not callable' errors

spillner committed:
BUGFIX: explicitly import packaging.version to provide version.parse

spillner committed:
Merge branch 'fixed-io' to resolve inconsistencies between Python 2.x and 3.x:
  - sort() now uses 3.x-style key() functions instead of 2.x-style cmp()
  - Converted file open() calls to use io.open() and specify file encoding
  - Added fallback encodings in write_to_pipe() (so non-ASCII characters
    can be sent to the pager correctly under Python 2.x)
[Rebased 14 Jan 2024 against upstream v52 master]

spillner committed:
Now caches package info files instead of having to re-download every time for packages not installed locally

spillner committed:
Further I/O cleanup: renamed write_to_pipe() as write_to_stream(), abstracted away read_from_file()/read_lines_from_file()

spillner committed:
Cleaned up and standardized Unicode/bytestring handling between Python 2 and Python 3

spillner committed:
Fix some issues with ChangeLog tracking (recognizing first duplicate entry) when switching back and forth between Python 2 and Python 3

spillner committed:
Protect isinstance(__, bytes) from being called under Python 2

spillner committed:
Remove obsolete URLopener() class to suppress warnings under Python 3
[Rebased 14 Jan 2024 against upstream v52 master]

spillner committed:
Fixed two unparenthesized print statements that were mis-merged in the rebase

spillner committed:
Re-added obsolete transient_cmp() for the purpose of passing the CI pipeline. Version bump to 53.

spillner committed:
Adjusted prioritised-pkgs tests to use transient_key() instead of transient_cmp()

spillner commented:
Noticed this had been sitting for about a year; I just rebased my independent version against upstream v52 and bumped the (proposed) version to 53 to account for the volume of new commands.  Passes all CI tests with the exception of an HttpError that looks like it may be an artifact of DNS resolution or security policies in the test environment (at any rate, downloads for "update"/"upgrade"/"info-not-installed"/etc. are working fine here).

Major changes include further Python 3 cleanup, adding some new command aliases, tweaking the behavior of "kernel-clean" to leave outdated packages on the system until the new repo version is actually installed, and a (versioned) local cache for .info files to avoid unnecessary network traffic.  The string/Unicode abstraction isn't aesthetic but it maximizes Python 2/3 code reuse and database interchangeability; hopefully it's something we never have to worry about again.

spillner committed:
Cleanup a few spots that were mishandled by the auto-merge tool during v52 rebase

spillner committed:
Corrected test harness to handle package states correctly.

aclemons commented:
There are a lot of different things going on in this one PR so I don't think I will be merging it as is.

I definitely won't be merging the aliases. I don't see any reason to clutter up the list of commands with duplicates. If someone is moving from an apt based system, there are already heaps of things to learn about how Slackware is different, and time will have to spent on how to manage packages. This is not even the official package manager for Slackware or anything, so someone has to learn this specific tool anyway.

I merged the .new prompt change and tagged a release with that just now. That had been hanging around for too long and that's on me.

I'll have a look at the info and upgradable commands. I might tease some of those out and add them.

Not sure how keen I on changing how kernel packages are handled. This is a significant change in how the tool works and will mean people might have to change their workflows.

I do still need to finish pulling in the remaining Python3 support.

spillner committed:
Fixed Py2 exception when launching slackroll with an invalid $PATH

spillner committed:
Replace packaging.version dependency with a custom comparison function that can handle version numbers not IAW PEP 440

spillner committed:
Marked regular expression as a raw string to avoid warnings in Python 3.11 and newer.

spillner committed:
Fixed up_to_date() check to recognize when a potential update is available in one remote repository, even if the currently installed version is still current in a different remote repository.  This is essential if you want to have some repositories (e.g. ktown) override the main Slackware repo.  Note, however, that it will make packages that have newer versions in extra/ or testing/ show up as 'transient' unless/until they are marked as frozen.

spillner committed:
Improved superseded-package determination to minimize false-positives when remote repos hold a version under multiple architectures, or use different build suffixes within the same architecture.

spillner committed:
Added --auto-select option to upgrade and upgrade-key-packages.
This reduces noise when you are combining repos that provide
alternate versions for common packages.

spillner committed:
Added the '--auto-select' option to list-upgrades as well. Changed the
semantics slightly, so that when multiple versions of a given package
are installed locally (should normally not be the case), --auto-select
will consider upgrades that match ANY of the locally installed suffixes.

spillner committed:
Support additional ChangeLog.txt files in third-party repos. Extended
database try_() interface to permit re-accessing open DBs, manual closure of
databases that are no longer needed, and higher-level exception handling
when desired (e.g. if failure to open a .db file is not necessarily fatal).
